lib/internal/fetch/headers.js

```diff
@@ -0,0 +1,212 @@
+          object[i + 1].toString('utf-8')
+        );
+      }
+    } else {
+      // all other array based entries
+      throw new ERR_INVALID_ARG_VALUE(
+        'init',
+        object,
+        'is not a valid array entry'
+      );
+    }
+  } else if (!isBoxedPrimitive(object)) {
+    // object of key/value entries
+    for (const { 0: name, 1: value } of ObjectEntries(object)) {
+      headers.append(name, value);
+    }
+  }
+}
+
+class Headers {
+  constructor(init) {
+    this[kHeadersList] = [];
+
+    if (init && validateObject(init, 'init', { allowArray: true }) === undefined) {
+      fill(this, init);
+    }
+  }
+
+  append(name, value) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+    const normalizedValue = normalizeAndValidateHeaderValue(name, value);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList][index + 1] += `, ${normalizedValue}`;
+    } else {
+      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue);
+    }
+  }
+
+  delete(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList].splice(index, 2);
+    }
+  }
+
+  get(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    if (this[kHeadersList][index] === normalizedName) {
+      return this[kHeadersList][index + 1];
+    }
+
+    return null;
+  }
+
+  has(name) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+
+    return this[kHeadersList][index] === normalizedName;
+  }
+
+  set(name, value) {
+    const normalizedName = normalizeAndValidateHeaderName(name);
+    const normalizedValue = normalizeAndValidateHeaderValue(name, value);
+
+    const index = binarySearch(this[kHeadersList], normalizedName);
+    if (this[kHeadersList][index] === normalizedName) {
+      this[kHeadersList][index + 1] = normalizedValue;
+    } else {
+      this[kHeadersList].splice(index, 0, normalizedName, normalizedValue);
+    }
+  }
+
+  *keys() {
+    for (const header of this) {
+      yield header[0];
+    }
+  }
+
+  *values() {
+    for (const header of this) {
+      yield header[1];
+    }
+  }
+
+  *entries() {
+    yield* this;
+  }
+
+  forEach(callback, thisArg) {
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      callback.call(
+        thisArg,
+        this[kHeadersList][index + 1],
+        this[kHeadersList][index],
+        this
+      );
+    }
+  }
+
+  *[SymbolIterator]() {
+    for (let index = 0; index < this[kHeadersList].length; index += 2) {
+      yield [this[kHeadersList][index], this[kHeadersList][index + 1]];
+    }
+  }
+}
```

Comment:
I think we should mimic how [`Map.prototype[@@iterator]`](https://tc39.es/ecma262/multipage/keyed-collections.html#sec-map.prototype-@@iterator) is defined:
```suggestion
}

Headers.prototype[SymbolIterator] = Headers.prototype.entries;
```